### PR TITLE
[LS] Add performance tests

### LINKED ISSUE #31920: [LS] Add performance tests
**Description:**
$subject

### PR DESCRIPTION
**Pull request by gayalkuruppu**
Fixes #31920 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionOpenDocPerformanceTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        position.setLine(positionObj.get("line").getAsInt());
+        position.setCharacter(positionObj.get("character").getAsInt());
+        JsonElement triggerCharElement = configJsonObject.get("triggerCharacter");
+        String triggerChar = triggerCharElement == null ? "" : triggerCharElement.getAsString();
+
+        long start = System.currentTimeMillis();
+        TestUtil.openDocument(serviceEndpoint, sourcePath);
+        long end = System.currentTimeMillis();
+        long responseTime = end - start;
+        responseString = TestUtil.getCompletionResponse(sourcePath.toString(), position,
```
[the remark this answers is not in the document]
Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionOpenDocPerformanceTest.java:68]: Should we pass the test even if the expected result and the response differ?

[the remark this answers is not in the document]
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionOpenDocPerformanceTest.java:68]: In this case yes. This is because the use-case we are going to test, is the project open. WDYT?

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/util/TestUtil.java [hunk 2]
```diff
@@ -695,4 +695,12 @@ public static String getResponseString(CompletableFuture<?> completableFuture) {
...
         return diagnostics;
     }
+
+    public static boolean isResponseWithinExpected(long responseTime) {
+        if (responseTime == 0) {
+            return false;
+        } else {
+            return responseTime < 3000;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/util/TestUtil.java:705]: No specific use case and changed to `Assert.assertEquals(responseTime < 3000, true);` by removing the method

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CodeActionPerformanceTest.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+            = BallerinaWorkspaceManager.getInstance(new LanguageServerContextImpl());
+    private static final LanguageServerContext serverContext = new LanguageServerContextImpl();
+
+    @BeforeClass
+    public void init() throws Exception {
+        this.serviceEndpoint = TestUtil.initializeLanguageSever();
+    }
+
+    @Test(dataProvider = "performance-data-provider")
+    public void testCodeAction(String config, String source) throws IOException, WorkspaceDocumentException {
```
[the remark this answers is not in the document]
Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CodeActionPerformanceTest.java:71]: Can you please elaborate?

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionPerformanceTest.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        JsonObject positionObj = configJsonObject.get("position").getAsJsonObject();
+        position.setLine(positionObj.get("line").getAsInt());
+        position.setCharacter(positionObj.get("character").getAsInt());
+        JsonElement triggerCharElement = configJsonObject.get("triggerCharacter");
+        String triggerChar = triggerCharElement == null ? "" : triggerCharElement.getAsString();
+
+        TestUtil.openDocument(serviceEndpoint, sourcePath);
+        long start = System.currentTimeMillis();
+        responseString = TestUtil.getCompletionResponse(sourcePath.toString(), position,
+                this.serviceEndpoint, triggerChar);
```
[the remark this answers is not in the document]
Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionPerformanceTest.java:116]: Removed the `getResponseTimeCompletion` method and returned a Pair<responseTime, responseString> inside `getResponseCompletion`

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionOpenDocPerformanceTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionOpenDocPerformanceTest.java:68]: +1

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CodeActionPerformanceTest.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                .filter(diag -> CommonUtil.isWithinRange(pos, diag.getRange()))
+                .collect(Collectors.toList());
+        CodeActionContext codeActionContext = new CodeActionContext(diags);
+
+        Range range = new Range(pos, pos);
+        long start = System.currentTimeMillis();
+        String res = TestUtil.getCodeActionResponse(serviceEndpoint, sourcePath.toString(), range, codeActionContext);
+        long end = System.currentTimeMillis();
+        long responseTime = end - start;
+        Assert.assertEquals(responseTime < 3000, true);
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CodeActionPerformanceTest.java:92]: When checking a condition, we can use `Assert.assertTrue(responseTime < 3000)` instead. Also, move `3000` to a class level constant (static variable)

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionOpenDocPerformanceTest.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+
+/**
+ * Test performance of openDocument in completions feature in language server.
+ */
+public class CompletionOpenDocPerformanceTest extends CompletionPerformanceTest {
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionOpenDocPerformanceTest.java:37]: I assume this test is to measure performance of `openDocument` call? If so, we should rename this to imply that. Or, we can include the `openDocument` performance tests in other tests (like completion) because we are going to perform the open document step anyway.

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CodeActionPerformanceTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (IMS94) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CodeActionPerformanceTest.java:70]: What @nadeeshaan means here is regarding code duplication. Since we have similar methods/snippets in existing code action and completion tests, we should try to reuse them by moving common logic into a utility class like `TestUtil`.

**Review by IMS94 — CHANGES_REQUESTED** [review]
Looks good for the 1st iteration. We can build on this. I have added some comments, let's address them as well.

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionOpenDocPerformanceTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionOpenDocPerformanceTest.java]: Removed

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionPerformanceTest.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+        TestUtil.openDocument(serviceEndpoint, sourcePath);
+        responseString = TestUtil.getCompletionResponse(sourcePath.toString(), position,
+                this.serviceEndpoint, triggerChar);
+        TestUtil.closeDocument(serviceEndpoint, sourcePath);
+
+        return responseString;
+    }
+
+    long getResponseTimeCompletion(JsonObject configJsonObject) throws IOException {
```
[the remark this answers is not in the document]
Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CompletionPerformanceTest.java]: `getResponseCompletion` has been added to the `testCompletion`. Hence we no longer need to return `Pair` as in the https://github.com/ballerina-platform/ballerina-lang/pull/32388/commits/db7e033402eb2ae006f5b5eb26bbd2f82b695a1a

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/CompletionTest.java [hunk 8]
```diff
@@ -86,21 +84,22 @@ public void test(String config, String configPath) throws WorkspaceDocumentExcep
         }
     }
 
-    String getResponse(JsonObject configJsonObject) throws IOException {
+    private String initializeSources(JsonObject configJsonObject) throws IOException {
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/CompletionTest.java:87]: Better if we can change the method name since the name does not reflect what it actually does.

File: language-server/modules/langserver-core/src/test/resources/performance/sources/main.bal [hunk 9]
```diff
@@ -0,0 +1,99 @@
+import wso2/nballerina.types as t;
+import wso2/nballerina.bir;
+import wso2/nballerina.front;
+import wso2/nballerina.nback;
+import wso2/nballerina.err;
+
+import ballerina/io;
+import ballerina/file;
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/test/resources/performance/sources/main.bal:8]: let's keep this file as a file with diagnostics and this source is good. If we need un resolved modules, let's use some other values other than the existing io, file and wso2 based imports.

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/CompletionTest.java (hunk 8 above)
Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/CompletionTest.java:87]: Shall we change the method name to another `getResponse`?

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/CompletionPerformanceTest.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    @Override
+    public String getResponse(Path sourcePath, Position position, String triggerChar) throws IOException {
+        TestUtil.openDocument(serviceEndpoint, sourcePath);
+        long start = System.currentTimeMillis();
+        String responseString = TestUtil.getCompletionResponse(sourcePath.toString(), position,
+                this.serviceEndpoint, triggerChar);
+        long end = System.currentTimeMillis();
+        TestUtil.closeDocument(serviceEndpoint, sourcePath);
+        long actualResponseTime = end - start;
+        int expectedResponseTime = Integer.parseInt(System.getProperty("responseTimeThreshold")) / 2;
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/CompletionPerformanceTest.java:53]: Same as codeaction scenario, let's move this to a util

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/CompletionTest.java [hunk 11]
```diff
@@ -49,7 +49,7 @@
  */
 public abstract class CompletionTest {
 
-    private Endpoint serviceEndpoint;
+    public Endpoint serviceEndpoint;
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/CompletionTest.java:52]: `protected`

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/hover/HoverProviderTest.java [hunk 12]
```diff
@@ -39,7 +39,7 @@
  * Test hover feature in language server.
  */
 public class HoverProviderTest {
-    private Endpoint serviceEndpoint;
+    public Endpoint serviceEndpoint;
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/hover/HoverProviderTest.java:42]: protected

File: language-server/modules/langserver-core/src/test/resources/performance/sources/projectls/main.bal [hunk 13]
```diff
@@ -0,0 +1,37 @@
+import projectls.module1;
+public function main() {
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/resources/performance/sources/projectls/main.bal:2]: Shall we format this document? Also the other bal files added.

File: tests/jballerina-integration-test/src/test/resources/observability/tracing_tests/Dependencies.toml [hunk 14]
```diff
@@ -7,5 +7,3 @@ version = "0.0.0"
 org = "ballerina"
 name = "observe"
 version = "0.9.0"
-
```
Comment (IMS94) [inline tests/jballerina-integration-test/src/test/resources/observability/tracing_tests/Dependencies.toml:10]: Unrelated change. Let's revert this.

**Review by IMS94 — CHANGES_REQUESTED** [review]
LGTM! Need some minor refactoring. Once done, we can merge.

File: tests/jballerina-integration-test/src/test/resources/observability/tracing_tests/Dependencies.toml (hunk 14 above)
Comment (gayalkuruppu) [reply tests/jballerina-integration-test/src/test/resources/observability/tracing_tests/Dependencies.toml:10]: [master](https://github.com/ballerina-platform/ballerina-lang/blame/master/tests/jballerina-integration-test/src/test/resources/observability/tracing_tests/Dependencies.toml) seem to have an extra line, hence keeping the change.

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/util/PerformanceTestUtils.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Test utils for running performance tests.
+ */
+public class PerformanceTestUtils {
+
+    /**
+     * Get response time threshold for Completion Performance Test.
+     */
+    public static int getCompletionResponseTimeThreshold() {
+        return Integer.parseInt(System.getProperty("responseTimeThreshold")) / 2;
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/util/PerformanceTestUtils.java:29]: Shall we move `"responseTimeThreshold"` to a `private static final String` variable defined in this class.

**Review by IMS94 — DISMISSED** [review]
LGTM! There's a minor comment. Please address that before merging.

**Review by IMS94 — APPROVED** [review]

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/AbstractCodeActionTest.java [hunk 16]
```diff
@@ -49,13 +49,14 @@
...
 
+
 /**
  * Test Cases for CodeActions.
  *
  * @since 2.0.0
  */
 public abstract class AbstractCodeActionTest {
-    private Endpoint serviceEndpoint;
+    public Endpoint serviceEndpoint;
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/AbstractCodeActionTest.java:59]: Shall we make this private

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/CodeActionPerformanceTest.java [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    @Test(dataProvider = "codeaction-data-provider")
+    public void test(String config, String source) throws IOException, WorkspaceDocumentException {
+        super.test(config, source);
+    }
+
+    @Override
+    public String getResponse(Path sourcePath, Range range, CodeActionContext codeActionContext) {
+        long start = System.currentTimeMillis();
+        String res = TestUtil.getCodeActionResponse(serviceEndpoint, sourcePath.toString(), range, codeActionContext);
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/CodeActionPerformanceTest.java:51]: We can call the super method as the the serviceEndpoint is not visible once it's made private.
```suggestion
        String res = super.getResponse(sourcePath.toString(), range, codeActionContext);
```

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/CompletionTest.java [hunk 18]
```diff
@@ -49,7 +49,7 @@
  */
 public abstract class CompletionTest {
 
-    private Endpoint serviceEndpoint;
+    protected Endpoint serviceEndpoint;
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/CompletionTest.java:52]: Sahll we make this private

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/CompletionPerformanceTest.java [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+    @Override
+    public void test(String config, String configPath) throws WorkspaceDocumentException, IOException {
+        super.test(config, configPath);
+    }
+
+    @Override
+    public String getResponse(Path sourcePath, Position position, String triggerChar) throws IOException {
+        TestUtil.openDocument(serviceEndpoint, sourcePath);
+        long start = System.currentTimeMillis();
+        String responseString = TestUtil.getCompletionResponse(sourcePath.toString(), position,
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/completion/CompletionPerformanceTest.java:49]: Shall we use the super method

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/hover/HoverProviderTest.java [hunk 20]
```diff
@@ -39,7 +39,7 @@
  * Test hover feature in language server.
  */
 public class HoverProviderTest {
-    private Endpoint serviceEndpoint;
+    protected Endpoint serviceEndpoint;
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/hover/HoverProviderTest.java:42]: Shall we make this private

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/hover/HoverPerformanceTest.java [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+
+    @Test(description = "Test Hover provider", dataProvider = "hover-data-provider")
+    public void testHover(String config) throws IOException {
+        super.testHover(config);
+    }
+
+    @Override
+    public String getResponse(Path sourcePath, Position position) {
+        long start = System.currentTimeMillis();
+        String responseString = parser.parse(TestUtil
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/hover/HoverPerformanceTest.java:45]: Sahll we use the super method

**Review by malinthar — CHANGES_REQUESTED** [review]

**Review by nadeeshaan — APPROVED** [review]
LGTM

**Review by malinthar — APPROVED** [review]
LGTM!

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/hover/HoverPerformanceTest.java (hunk 21 above)
Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/hover/HoverPerformanceTest.java:45]: leaving this change to keep the implementation consistent
